File: src/Api/Services/PostGenerationService.cs
using System.Net.Http;
using System.Text.Json;
using Azure;
using Microsoft.Extensions.Options;
using OpenAI.Chat;
using PostGenerator.Api.Options;

namespace PostGenerator.Api.Services;

public class PostGenerationService : IPostGenerationService
{
    private readonly IAzureOpenAIClientProvider _clientProvider;
    private readonly AzureOpenAIOptions _options;
    private readonly ILogger<PostGenerationService> _logger;

    public PostGenerationService(
        IAzureOpenAIClientProvider clientProvider,
        IOptions<AzureOpenAIOptions> options,
        ILogger<PostGenerationService> logger)
    {
        _clientProvider = clientProvider;
        _options = options.Value;
        _logger = logger;
    }

    public async Task<IReadOnlyList<GeneratedPostItem>> GeneratePostsAsync(GeneratePostOptions options, CancellationToken cancellationToken = default)
    {
        var platformNote = options.Platform.Equals("TikTok", StringComparison.OrdinalIgnoreCase) && options.TikTokScriptDurationSeconds.HasValue
            ? $" Each post must include a script suitable for a {options.TikTokScriptDurationSeconds} second video (approx {options.TikTokScriptDurationSeconds * 2} words)."
            : "";

        var systemPrompt = "You are a social media content writer. Generate posts as JSON. For each post return exactly: \"content\" (the post text), \"script\" (only for TikTok, the video script), \"hashtags\" (JSON array of hashtag strings). Be concise and match the requested tone and length.";
        var userPrompt = $"Generate {options.NumPosts} {(options.Linked ? "linked" : "standalone")} posts for platform: {options.Platform}. Topic: {options.TopicDetail}. Tone: {options.Tone ?? "professional"}. Length: {options.Length ?? "medium"}.{platformNote} Return a JSON array of objects, each with \"content\", \"script\" (optional), \"hashtags\" (array of strings). No markdown, only the raw JSON array.";

        var content = await CallChatCompletionAsync(systemPrompt, userPrompt, 4000, cancellationToken);
        if (content == null) return Array.Empty<GeneratedPostItem>();

        return ParseContentAsPostItems(content);
    }

    public async Task<GeneratedPostItem?> GenerateSinglePostAsync(GeneratePostOptions options, int index, IReadOnlyList<string>? previousContents, CancellationToken cancellationToken = default)
    {
        var platformNote = options.Platform.Equals("TikTok", StringComparison.OrdinalIgnoreCase) && options.TikTokScriptDurationSeconds.HasValue
            ? $" The post must include a script suitable for a {options.TikTokScriptDurationSeconds} second video (approx {options.TikTokScriptDurationSeconds * 2} words)."
            : "";

        var linkedContext = options.Linked && previousContents is { Count: > 0 }
            ? $" Previous posts in this series (for continuity): {string.Join(" | ", previousContents)}. Write the next post that follows naturally."
            : "";

        var systemPrompt = "You are a social media content writer. Generate a single post as JSON. Return exactly one object with: \"content\" (the post text), \"script\" (only for TikTok, the video script), \"hashtags\" (JSON array of hashtag strings). Be concise and match the requested tone and length. No markdown, only the raw JSON object.";
        var userPrompt = $"Generate post {index} of {options.NumPosts} for platform: {options.Platform}. Topic: {options.TopicDetail}. Tone: {options.Tone ?? "professional"}. Length: {options.Length ?? "medium"}.{platformNote}{linkedContext} Return a single JSON object with \"content\", \"script\" (optional), \"hashtags\" (array of strings).";

        var content = await CallChatCompletionAsync(systemPrompt, userPrompt, 2000, cancellationToken);
        if (content == null) return null;

        return ParseContentAsSinglePostItem(content);
    }

    private async Task<string?> CallChatCompletionAsync(string systemPrompt, string userPrompt, int maxTokens, CancellationToken cancellationToken)
    {
        var azureClient = _clientProvider.GetChatClient();
        if (azureClient == null)
            return null;

        var chatClient = azureClient.GetChatClient(_options.ChatDeploymentName);
        var messages = new ChatMessage[]
        {
            new SystemChatMessage(systemPrompt),
            new UserChatMessage(userPrompt),
        };
        var completionOptions = new ChatCompletionOptions
        {
            MaxOutputTokenCount = maxTokens,
            Temperature = 0.7f,
        };

        try
        {
            var completion = await chatClient.CompleteChatAsync(messages, completionOptions, cancellationToken);
            var content = completion.Value.Content;
            if (content == null || content.Count == 0)
                return null;
            var result = string.Concat(content
                .Where(p => p.Kind == ChatMessageContentPartKind.Text && p.Text != null)
                .Select(p => p.Text));
            return StripMarkdown(result.Trim());
        }
        catch (RequestFailedException ex)
        {
            throw new HttpRequestException($"Azure OpenAI returned {ex.Status} {ex.Message}. {ex.ErrorCode}", ex);
        }
    }

    private static string StripMarkdown(string content)
    {
        if (content.StartsWith("```"))
            content = content.Replace("```json", "").Replace("```", "").Trim();
        return content;
    }

    private List<GeneratedPostItem> ParseContentAsPostItems(string content)
    {
        try
        {
            var items = JsonSerializer.Deserialize<JsonElement[]>(content);
            if (items == null) return new List<GeneratedPostItem>();
            var list = new List<GeneratedPostItem>();
            foreach (var item in items)
                list.Add(ParsePostItem(item));
            return list;
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to parse Azure OpenAI response as JSON array, using raw content as single post");
            return new List<GeneratedPostItem> { new GeneratedPostItem(content, null, null) };
        }
    }

    private GeneratedPostItem ParseContentAsSinglePostItem(string content)
    {
        try
        {
            var item = JsonSerializer.Deserialize<JsonElement>(content);
            return ParsePostItem(item);
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to parse Azure OpenAI response as JSON object, using raw content");
            return new GeneratedPostItem(content, null, null);
        }
    }

    private static GeneratedPostItem ParsePostItem(JsonElement item)
    {
        var postContent = item.TryGetProperty("content", out var c) ? c.GetString() ?? "" : "";
        var script = item.TryGetProperty("script", out var s) ? s.GetString() : null;
        string? hashtagsJson = null;
        if (item.TryGetProperty("hashtags", out var h) && h.ValueKind == JsonValueKind.Array)
            hashtagsJson = h.GetRawText();
        return new GeneratedPostItem(postContent, script, hashtagsJson);
    }
}

File: src/lib/services/post-generation-service.ts
import { complete } from "./chat-completion";

export interface GeneratePostOptions {
  topicDetail: string;
  numPosts: number;
  platform: string;
  linked: boolean;
  tone?: string | null;
  length?: string | null;
  tiktokScriptDurationSeconds?: number | null;
}

export interface GeneratedPostItem {
  content: string;
  script?: string | null;
  hashtagsJson?: string | null;
}

function parsePostItems(content: string): GeneratedPostItem[] {
  try {
    const parsed = JSON.parse(content);
    const arr = Array.isArray(parsed) ? parsed : [parsed];
    return arr.map((item: Record<string, unknown>) => {
      const content = typeof item.content === "string" ? item.content : "";
      const script =
        typeof item.script === "string" ? item.script : null;
      let hashtagsJson: string | null = null;
      if (Array.isArray(item.hashtags)) {
        hashtagsJson = JSON.stringify(item.hashtags);
      }
      return { content, script, hashtagsJson };
    });
  } catch {
    return [{ content, script: null, hashtagsJson: null }];
  }
}

function parseSinglePostItem(content: string): GeneratedPostItem {
  try {
    const item = JSON.parse(content) as Record<string, unknown>;
    const c = typeof item.content === "string" ? item.content : content;
    const script = typeof item.script === "string" ? item.script : null;
    let hashtagsJson: string | null = null;
    if (Array.isArray(item.hashtags)) {
      hashtagsJson = JSON.stringify(item.hashtags);
    }
    return { content: c, script, hashtagsJson };
  } catch {
    return { content, script: null, hashtagsJson: null };
  }
}

export async function generatePosts(
  options: GeneratePostOptions
): Promise<GeneratedPostItem[]> {
  const platformNote =
    options.platform.toLowerCase() === "tiktok" && options.tiktokScriptDurationSeconds
      ? ` Each post must include a script suitable for a ${options.tiktokScriptDurationSeconds} second video (approx ${options.tiktokScriptDurationSeconds * 2} words).`
      : "";

  const systemPrompt =
    "You are a social media content writer. Generate posts as JSON. For each post return exactly: \"content\" (the post text), \"script\" (only for TikTok, the video script), \"hashtags\" (JSON array of hashtag strings). Be concise and match the requested tone and length.";
  const userPrompt = `Generate ${options.numPosts} ${options.linked ? "linked" : "standalone"} posts for platform: ${options.platform}. Topic: ${options.topicDetail}. Tone: ${options.tone ?? "professional"}. Length: ${options.length ?? "medium"}.${platformNote} Return a JSON array of objects, each with "content", "script" (optional), "hashtags" (array of strings). No markdown, only the raw JSON array.`;

  const content = await complete(systemPrompt, userPrompt, 4000);
  if (!content) return [];

  return parsePostItems(content);
}

export async function generateSinglePost(
  options: GeneratePostOptions,
  index: number,
  previousContents: string[]
): Promise<GeneratedPostItem> {
  const platformNote =
    options.platform.toLowerCase() === "tiktok" && options.tiktokScriptDurationSeconds
      ? ` The post must include a script suitable for a ${options.tiktokScriptDurationSeconds} second video (approx ${options.tiktokScriptDurationSeconds * 2} words).`
      : "";

  const linkedContext =
    options.linked && previousContents.length > 0
      ? ` Previous posts in this series (for continuity): ${previousContents.join(" | ")}. Write the next post that follows naturally.`
      : "";

  const systemPrompt =
    "You are a social media content writer. Generate a single post as JSON. Return exactly one object with: \"content\" (the post text), \"script\" (only for TikTok, the video script), \"hashtags\" (JSON array of hashtag strings). Be concise and match the requested tone and length. No markdown, only the raw JSON object.";
  const userPrompt = `Generate post ${index} of ${options.numPosts} for platform: ${options.platform}. Topic: ${options.topicDetail}. Tone: ${options.tone ?? "professional"}. Length: ${options.length ?? "medium"}.${platformNote}${linkedContext} Return a single JSON object with "content", "script" (optional), "hashtags" (array of strings).`;

  const content = await complete(systemPrompt, userPrompt, 2000);
  if (!content) throw new Error("Post generation returned empty");

  return parseSinglePostItem(content);
}
